connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java

```diff
@@ -544,6 +555,95 @@ public int generation() {
+        log.trace("Submitting validate configuration request {}", connType);
+
+        addRequest(
+            new Callable<Void>() {
+                @Override
+                public Void call() throws Exception {
+                    ConnectorConfig connConfig = new ConnectorConfig(connectorConfig);
+                    String connName = connConfig.getString(ConnectorConfig.NAME_CONFIG);
+                    ConfigDef configDef = getConfig(connType);
+                    Connector connector = getConnector(connType, connName);
+                    List<Config> configs = connector.validate(configDef, connectorConfig);
+                    List<ConfigInfo> configInfoList = new LinkedList<>();
+                    int errorCount = 0;
+                    for (Config config : configs) {
+                        ConfigInfo configInfo = new ConfigInfo(config.getName(), config
+                            .getValue(), config.getRecommendedValues(), config.getErrorMessages(), config.isVisible());
+                        configInfoList.add(configInfo);
+                        errorCount += config.getErrorMessages().size();
+                    }
+                    ConfigInfos
+                        configInfos = new ConfigInfos(connName, errorCount, configInfoList);
+                    callback.onCompletion(null, configInfos);
+                    return null;
+                }
+            },
+            forwardErrorCallback(callback)
+        );
+
+    }
+
+    private ConfigDef getConfig(String connType) {
+        if (configs.containsKey(connType)) {
+            return configs.get(connType);
+        } else {
+            ConfigDef configDef = worker.getConnectorConfigDef(connType);
+            configs.put(connType, configDef);
+            return configDef;
+        }
+    }
+
+    private Connector getConnector(String connType, String connName) {
+        if (tempConnectors.containsKey(connName)) {
+            return tempConnectors.get(connName);
+        } else {
+            Connector connector = worker.getConnector(connType);
+            tempConnectors.put(connName, connector);
+            return connector;
+        }
+    }
+
+    private ConfigKeyInfo convertConfigKey(ConfigDef.ConfigKey configKey) {
+        String name = configKey.name;
+        String type = configKey.type.name();
+        Object defaultValue = configKey.defaultValue;
+        String importance = configKey.importance.name();
+        String documentation = configKey.documentation;
+        String group = configKey.group;
+        int orderInGroup = configKey.orderInGroup;
+        String width = configKey.width.name();
+        String displayName = configKey.displayName;
+        List<String> dependents = configKey.dependents;
+        return new ConfigKeyInfo(name, type, defaultValue, importance, documentation, group, orderInGroup, width, displayName, dependents);
```

Comment:
The fact that these are quite similar makes me wonder if we can use the same object for both `ConfigKey` and `ConfigKeyInfo`? A lot seems like it could be handled just by the right annotations. Any reason we're not doing this? Especially if `ConfigKey` became public in `ConfigDef`?